sema: analyze struct field bodies in a second pass, to allow them to use the layout of the struct itself

kcbanner opened the pull request:
This PR is based off of https://github.com/ziglang/zig/pull/17658.

The main goal of this PR was to resolve the circular dependency error you used to get when doing something like this:

```zig
const StructInner = extern struct {
    outer: StructOuter = std.mem.zeroes(StructOuter),
};

const StructMiddle = extern struct {
    outer: ?*StructInner,
    inner: ?*StructOuter,
};

const StructOuter = extern struct {
    middle: StructMiddle = std.mem.zeroes(StructMiddle),
};

test "circular dependency through pointer field of a struct" {
    var outer: StructOuter = .{};
    try expect(outer.middle.outer == null);
    try expect(outer.middle.inner == null);
}
```

```
test\behavior\container_circular_dependency.zig:28:29: error: struct 'container_circular_dependency.StructMiddle' depends on itself
const StructMiddle = extern struct {
                     ~~~~~~~^~~~~~
test\behavior\container_circular_dependency.zig:34:5: note: while checking this field
    middle: StructMiddle = std.mem.zeroes(StructMiddle),
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This is accomplished by breaking out the struct field init body resolution into a separate pass, so it can happen after the struct's layout has been resolved. 

`semaStructFields` now caches the ranges of the init bodies in the `StructType` trailing data. The init bodies are then resolved by `resolveStructFieldInits`, which is called before the inits are actually required. Within the init bodies, the struct decl's instruction is repurposed to refer to the field type itself. This is to allow us to easily rebuild the `inst_map` mapping required for the init body instructions to refer to the field type. Thanks to @mlugg for the guidance on this one!

An additional benefit of this change is that we no longer need to make incorrect guesses about struct alignment when a field init wants to use its container's alignment. Previously, this test would fail because the struct alignment would be guessed to be pointer-aligned:

```zig
test "initializer uses own alignment" {
    const S = struct {
        x: u32 = @alignOf(@This()) + 1,
    };

    var s: S = .{};
    try expectEqual(4, @alignOf(S));
    try expectEqual(@as(usize, 5), s.x);
}
```

I updated the alignment guessing logic to mark if the guess was made, and it's checked in `resolveStructLayout`. I added a similar check for unions as well. 

I checked to see if the guessing logic could be removed entirely, but there are some cases where it's still required:

```zig
test "struct field has a pointer to an aligned version of itself" {
    const E = struct {
        next: *align(1) @This(),
    };
    var e: E = undefined;
    e = .{ .next = &e };

    try expect(&e == e.next);
}
```

I may look into how to resolve that in a future PR.

Additional changes:
- Autodoc fixup to handle the repurposes instruction in init bodies
- CBE and wasm fixes for handling union representations that have unknown tags. Some of the new test cases exposed this problem that was introduced in https://github.com/ziglang/zig/pull/17352.

Vexu commented:
Fixes #14005

kcbanner commented:
> Fixes #14005

I added this as a test case and it revealed an unnecessary call I was making to `resolveStructFieldInits` in `resolveStructFully`, thanks!

mlugg commented on src/InternPool.zig:
```diff
@@ -400,6 +401,23 @@ pub const Key = union(enum) {
+                len: u32,
+
+                pub fn get(this: @This(), ip: *const InternPool) []InitBody {
+                    // Workaround for not being able to case between slices of different lengths
```
s/case/cast

mlugg commented on src/InternPool.zig:
```diff
@@ -497,6 +516,14 @@ pub const Key = union(enum) {
         }
 
+        /// The returned pointer expires with any addition to the `InternPool`.
+        /// Asserts the is not packed.
```
```suggestion
        /// Asserts that the struct is packed.
```

mlugg commented on src/InternPool.zig:
```diff
@@ -546,6 +573,28 @@ pub const Key = union(enum) {
 
+        pub fn setInitsWip(s: @This(), ip: *InternPool) bool {
+            switch (s.layout) {
+                inline else => |layout| {
```
Why use `inline else` here? The `Packed` case is the only one that needs special handling, we don't actually depend on it being comptime-known, and the rest of the body is trivial so it's fine to just duplicate it:
```zig
switch (s.layout) {
    .Packed => {
        const flag = &s.packedFlagsPtr(ip).field_inits_wip;
        if (flag.*) return true;
        flag.* = true;
        return false;
    },
    .Auto, .Extern => {
        const flag = &s.flagsPtr(ip).field_inits_wip;
        if (flag.*) return true;
        flag.* = true;
        return false;
    },
}
```

mlugg commented on src/InternPool.zig:
```diff
@@ -588,6 +637,20 @@ pub const Key = union(enum) {
+        }
+
+        pub fn setHaveFieldInits(s: @This(), ip: *InternPool) void {
+            return switch (s.layout) {
```
No need for this `return`.

mlugg commented on src/InternPool.zig:
```diff
@@ -2985,14 +3048,25 @@ pub const Tag = enum(u8) {
-    /// 2. init: Index for each fields_len // if tag is type_struct_packed_inits
+    ///
+    /// if tag is type_struct_packed_inits:
+    /// 2. body_indices: Zir.Inst.Index // for each field in declared order, a pair of start index and length.
```
This information is redundant, and should not be duplicated in the InternPool. Instead, it should be re-fetched from ZIR extra data when resolving field inits. Duplicating this information in the pool is particularly undesirable as it may change across incremental updates.

Merge blocker.

kcbanner replied:
Ah, I hadn't considered incremental compilation. I'll duplicate the logic from `semaStructFields` that works through the ZIR extra data instead.

kcbanner replied:
I've reworked this to re-fetch the init body ranges. I moved some common preamble logic into `semaStructFieldsLen`.

mlugg commented on src/InternPool.zig:
```diff
@@ -3015,6 +3089,7 @@ pub const Tag = enum(u8) {
     ///    names_map: MapIndex,
     ///    name: NullTerminatedString // for each field in declared order
     /// 2. if any_default_inits:
+    ///    body_indices: Zir.Inst.Index // for each field in declared order, a pair of start index and length.
```
Same as above.

mlugg commented on test/behavior/container_circular_dependency.zig:
```diff
@@ -0,0 +1,41 @@
+const std = @import("std");
```
Generally, try to avoid making overly specific files in `test/behavior/`. These would be better suited to `union.zig` and `struct.zig` respectively.

Also worth noting: this file isn't actually included in `test/behavior.zig` currently!

kcbanner replied:
Moved these into struct.zig / union.zig. I had originally made them their own file because they had to be in the root namespace to compile (otherwise `StructOuter` is an undefined identifier) and I didn't want to pollute the root namespace.

kcbanner commented:
I've added a fix for the CBE not handling union representations with an unknown tag - this was broken since https://github.com/ziglang/zig/pull/17352.

mlugg commented:
As a note, you accidentally left in both `Add compile error test case for...` commits from the other branch, but that can be sorted during merge no problem

mlugg commented on src/codegen/c.zig:
```diff
@@ -1499,56 +1499,84 @@ pub const DeclGen = struct {
+                            try writer.writeAll("*((");
                             try dg.renderType(writer, ty);
-                            try writer.writeByte(')');
+                            try writer.writeAll("*)(");
```
I'm not entirely sure how we render more complicated C types, so maybe it works for us, but just tacking a `*` on the end of the type isn't generally correct. I'd prefer if this constructed the actual pointer type (`mod.singleConstPtrType` should work) and rendered that.

kcbanner replied:
Good point, done 👍

mlugg commented on src/Sema.zig:
```diff
@@ -35641,24 +35546,18 @@ fn resolveInferredErrorSetTy(
-    const namespace_index = struct_type.namespace.unwrap() orelse decl.src_namespace;
-    const zir = mod.namespacePtr(namespace_index).file_scope.zir;
-    const zir_index = struct_type.zir_index;
+fn semaStructFieldsLen(zir: Zir, zir_index: Zir.Inst.Index) struct {
```
Functions beginning `sema` generally begin semantic analysis of something (e.g. `semaDecl` analyzes a `Decl`, `semaStructFields` analyzes the field types of a struct). This should be something like `structZirInfo` instead.

kcbanner replied:
Renamed.

mlugg commented on src/Sema.zig:
```diff
@@ -35941,20 +35858,158 @@ fn semaStructFields(
+            struct_type.field_inits.get(ip)[field_i] = field_init;
         }
-    } else {
-        struct_type.setHaveFieldInits(ip);
```
Hm, why is this call gone?

kcbanner replied:
It's still there - the diff for that commit made it a bit confusing since I moved a function:

`if (!any_inits) struct_type.setHaveFieldInits(ip);`

kcbanner commented:
> As a note, you accidentally left in both `Add compile error test case for...` commits from the other branch, but that can be sorted during merge no problem
> 
> EDIT: ah, you removed it in a later commit - no worries, if it's okay with you @kcbanner I'm gonna force-push to your branch in a moment to reshuffle a few commits, but no material changes

I've addressed your feedback and made a couple fixes for big-endian arches. 

Feel free to shuffle commits, thanks!

kcbanner commented:
Rebased to resolve conflicts.

kcbanner commented:
Fixes https://github.com/ziglang/zig/issues/17563

mlugg reviewed (approved):
Sorry for how long it took me to get around to this - it all looks good. I'm going to do one final rebase (and shuffle a few commits around), then I'll set this to auto-merge. Thanks!

andrewrk commented:
This broke #17892. I haven't tracked it down yet but I have a stack trace of optimized CBE: https://github.com/ziglang/zig/pull/17892#issuecomment-1801004192

Also I don't like those changes to packed struct representation because I don't think a `flags` field needs to be added to packed struct. I avoided it so far, and now 30 bits are going to be wasted for 2, which feels bad.

mlugg commented:
Good point on the packed struct flags - I'm not sure if we can eliminate those bits entirely, but if not then I'd suggest we could perhaps repurpose a few bits of `fields_len` in `TypeStructPacked`.

linusg commented:
This caused a false `struct depends on itself` in [kiesel](https://codeberg.org/kiesel-js/kiesel/) - unfortunately the project is too large for me to easily reduce this, but building with zig master should reproduce:

```
src/types/language/Object/InternalMethods.zig:1:1: error: struct 'types.language.Object.InternalMethods' depends on itself
//! 6.1.7.2 Object Internal Methods and Internal Slots
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

kcbanner commented:
> Good point on the packed struct flags - I'm not sure if we can eliminate those bits entirely, but if not then I'd suggest we could perhaps repurpose a few bits of `fields_len` in `TypeStructPacked`.

If we can steal those two bits from `fields_len`, I'm happy to PR that.
